### ISSUE TITLE
appending to MODULEPATH is not idempotent in subshells and makes Lmod complain

### ISSUE THREAD

**Issue by rkdarst** [CSC-IT-Center-for-Science/ansible-role-fgci-bash#5]
This issue comes from Aalto:

In templates/modules-cvmfs.sh, the path is unconditionally appended to `$MODULEPATH`.  User bashrc sources /etc/bashrc which sources /etc/profile.d/*.  So subshells re-append the same things to `$MODULEPATH`.  (the previous at least on centos7)  Lmod then notices that the paths have changed (though trivially) and complains.  This means, for example, if you start tmux then it breaks modules, and same anywhere subshells are started.

My own solution for PATH idempotence for any path variable (but there are countless others that may be better):

```
PATHADD2 () { if eval [ x"\$$1" = x'' ] ; then eval "export $1=$2" ; else eval "echo \"\$$1\"" | grep "$2\(:\|$\)" 2>&1 > /dev/null || eval "export $1=\$$1:$2" ; fi }

PATHADD2 MODULEPATH /share/apps/easybuild/modules/all
PATHADD2 MODULEPATH /share/apps/modules
PATHADD2 MODULEPATH /cvmfs/fgi.csc.fi/modules/el7/all

unset -f PATHADD2
```

I use bash, I'm not sure about the portability but my initial tests show it should work on zsh and dash at least.

The .csh template should also be edited, but I haven't worked on that yet.  (However maybe csh doesn't have the rc/profile loop that will cause the problems).

Thoughts?

**Comment by rkdarst** [comment]
Er, of course my paste needs to be adapted to use the jinja templates... it's probably simplest to make a new function that splits the colon-separated input then individually appends them.

**Comment by rkdarst** [comment]
The repo [rkdarst/ansible-role-fgci-bash](https://github.com/rkdarst/ansible-role-fgci-bash/commits/master) has a fix as it's top two commits.  I've tested with tcsh,bash,zsh,dash and deployed it to Aalto's login node.  I will make a pull request in a few days after testing confirms it works.

The change is overly complicated to be portable.  The sh-based ones are easy enough because I copied my own PATHADD2 function, but the loop had to be changed to not have a bashism.

The csh one was difficult since I don't know csh, so someone that does know it could probably improve it and probably should take a look.  It is also less general since csh doesn't have functions(!).
